File: typescript/ts_sdk/src/wns/resolver.ts
import {
  HandleGoneError,
  HandleMovedError,
  HandleNotFoundError,
  HandleResolutionError,
} from '../errors/index.js';
import type { HandleResolutionDocument, ResolveHandleOptions } from './types.js';
import { HandleStatus } from './types.js';
import { buildResolutionUrl, parseWbaUri, validateHandle } from './validator.js';

export async function resolveHandle(
  handle: string,
  options: ResolveHandleOptions = {}
): Promise<HandleResolutionDocument> {
  const bareHandle = stripWbaScheme(handle);
  const [localPart, domain] = validateHandle(bareHandle);
  const normalized = `${localPart}.${domain}`;
  const baseUrl = options.baseUrlOverride?.replace(/\/$/, '');
  const url = baseUrl
    ? `${baseUrl}/.well-known/handle/${localPart}`
    : buildResolutionUrl(localPart, domain);

  void options.verifySsl;
  const timeoutMs = Math.round((options.timeoutSeconds ?? 10) * 1000);
  const response = await fetch(url, {
    headers: { Accept: 'application/json' },
    redirect: 'manual',
    signal: AbortSignal.timeout(timeoutMs),
  }).catch((error) => {
    throw new HandleResolutionError(
      `Network error resolving handle '${normalized}': ${(error as Error).message}`,
      502,
      error as Error
    );
  });

  if (response.status === 301) {
    throw new HandleMovedError(
      `Handle '${normalized}' has been migrated`,
      response.headers.get('Location') ?? ''
    );
  }
  if (response.status === 404) {
    throw new HandleNotFoundError(`Handle '${normalized}' does not exist`);
  }
  if (response.status === 410) {
    throw new HandleGoneError(`Handle '${normalized}' has been permanently revoked`);
  }
  if (!response.ok) {
    throw new HandleResolutionError(
      `Unexpected status ${response.status} resolving '${normalized}'`,
      502
    );
  }

  const payload = (await response.json()) as Record<string, unknown>;
  const document: HandleResolutionDocument = {
    handle: String(payload.handle ?? ''),
    did: String(payload.did ?? ''),
    status: normalizeStatus(String(payload.status ?? '')),
    updated: payload.updated ? String(payload.updated) : undefined,
  };
  if (document.handle.toLowerCase() !== normalized) {
    throw new HandleResolutionError(
      `Handle mismatch: requested '${normalized}', got '${document.handle}'`,
      502
    );
  }
  return document;
}

export async function resolveHandleFromUri(
  wbaUri: string,
  options: ResolveHandleOptions = {}
): Promise<HandleResolutionDocument> {
  const parsed = parseWbaUri(wbaUri);
  return resolveHandle(parsed.handle, options);
}

function stripWbaScheme(handleOrUri: string): string {
  return handleOrUri.startsWith('wba://') ? handleOrUri.slice('wba://'.length) : handleOrUri;
}

function normalizeStatus(value: string): HandleStatus {
  switch (value.toLowerCase()) {
    case HandleStatus.Active:
      return HandleStatus.Active;
    case HandleStatus.Suspended:
      return HandleStatus.Suspended;
    case HandleStatus.Revoked:
      return HandleStatus.Revoked;
    default:
      throw new HandleResolutionError(`Unexpected handle status '${value}'`, 502);
  }
}

File: dart/lib/src/wns/resolver.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import '../errors.dart';
import 'types.dart';
import 'validator.dart';

Future<HandleResolutionDocument> resolveHandle(
  String handle, {
  ResolveHandleOptions options = const ResolveHandleOptions(),
  http.Client? client,
}) async {
  final parsed = validateHandle(handle.replaceFirst(RegExp(r'^wba://'), ''));
  final normalized = parsed.handle;
  final url = options.baseUrlOverride == null
      ? buildResolutionUrl(parsed.localPart, parsed.domain)
      : '${options.baseUrlOverride!.replaceFirst(RegExp(r'/+$'), '')}/.well-known/handle/${parsed.localPart}';
  final owned = client == null;
  final httpClient = client ?? http.Client();
  try {
    final response = await httpClient
        .get(Uri.parse(url), headers: const {'Accept': 'application/json'})
        .timeout(options.timeout);
    switch (response.statusCode) {
      case 200:
        break;
      case 301:
        throw AnpWnsException("handle '$normalized' has been migrated");
      case 404:
        throw AnpWnsException("handle '$normalized' does not exist");
      case 410:
        throw AnpWnsException(
          "handle '$normalized' has been permanently revoked",
        );
      default:
        throw AnpWnsException(
          "unexpected status ${response.statusCode} resolving '$normalized': ${response.body}",
        );
    }
    final decoded = jsonDecode(response.body);
    if (decoded is! Map) {
      throw const AnpWnsException('resolution document is not an object');
    }
    final map = Map<String, Object?>.from(decoded.cast<String, Object?>());
    final responseHandle =
        map['handle']?.toString().toLowerCase() ?? normalized;
    if (responseHandle != normalized) {
      throw AnpWnsException(
        "handle mismatch: requested '$normalized', got '$responseHandle'",
      );
    }
    return HandleResolutionDocument(
      handle: normalized,
      did: map['did']?.toString() ?? '',
      status: HandleStatus.values.firstWhere(
        (s) => s.name == (map['status'] ?? 'active'),
        orElse: () => HandleStatus.active,
      ),
      didDocument: map['didDocument'] is Map
          ? Map<String, Object?>.from(
              (map['didDocument'] as Map).cast<String, Object?>(),
            )
          : null,
    );
  } finally {
    if (owned) httpClient.close();
  }
}

Future<HandleResolutionDocument> resolveHandleFromUri(
  String uri, {
  ResolveHandleOptions options = const ResolveHandleOptions(),
  http.Client? client,
}) => resolveHandle(parseWbaUri(uri).handle, options: options, client: client);
